### PR TITLE
Mochitest for payment request construction.

### PR DESCRIPTION
**Pull request by edenchuang**

### CODE REVIEW

File: dom/payments/test/browser_payment_construction.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+      let paymentSrv = Cc["@mozilla.org/dom/payments/payment-request-service;1"].getService(Ci.nsIPaymentRequestService);
+      ok(paymentSrv, "Get payment request service failed.");
+      let paymentEnum = paymentSrv.enumerate();
+      ok(paymentEnum.hasMoreElements(), "Should have at least one request in nsIPaymentRequestService.");
+      while (paymentEnum.hasMoreElements()) {
+        let payment = paymentEnum.getNext().QueryInterface(Ci.nsIPaymentRequest);
+        ok(payment, "Should get existing payment.");
+
+        // checking the passed PaymentMethods parameter
+        is(payment.paymentMethods.length, 1, "Should only one supported payment method.");
```
Comment (bentian) [inline dom/payments/test/browser_payment_construction.js]: Revise to `Should support only one payment method`.

File: dom/payments/test/browser_payment_construction.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+add_task(function*() {
+  yield BrowserTestUtils.withNewTab(kTestPage,
+    function*(browser) {
+      let paymentSrv = Cc["@mozilla.org/dom/payments/payment-request-service;1"].getService(Ci.nsIPaymentRequestService);
+      ok(paymentSrv, "Get payment request service failed.");
+      let paymentEnum = paymentSrv.enumerate();
+      ok(paymentEnum.hasMoreElements(), "Should have at least one request in nsIPaymentRequestService.");
+      while (paymentEnum.hasMoreElements()) {
+        let payment = paymentEnum.getNext().QueryInterface(Ci.nsIPaymentRequest);
+        ok(payment, "Should get existing payment.");
```
Comment (bentian) [inline dom/payments/test/browser_payment_construction.js]: `Failed to get existing payment request`.

File: dom/payments/test/browser_payment_construction.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        is(details.id, "test details", "details.id should be 'test details'.");
+        is(details.totalItem.label, "Donation", "total item's label should be 'Donation'.");
+        is(details.totalItem.amount.currency, "USD", "total item's currency should be 'USD'.");
+        is(details.totalItem.amount.value, "55.00", "total item's value should be '55.00'.");
+        let displayItems = details.displayItems;
+        is(displayItems.length, 2, "displayItems' length should be 2.");
+        let item = displayItems.queryElementAt(0, Ci.nsIPaymentItem);
+        is(item.label, "Original donation amount", "item's label should be 'Original donation amount'.");
+        is(item.amount.currency, "USD", "item's currency should be 'USD'.");
+        is(item.amount.value, "-65.00", "item's value should be '-65.00'.");
```
Comment (bentian) [inline dom/payments/test/browser_payment_construction.js]: `1st item's ...` for above.

File: dom/payments/test/browser_payment_construction.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        item = displayItems.queryElementAt(1, Ci.nsIPaymentItem);
+        is(item.label, "Friends and family discount", "item's label should be 'Friends and family discount'.");
+        is(item.amount.currency, "USD", "item's currency should be 'USD'.");
+        is(item.amount.value, "10.00", "item's value should be '10.00'.");
```
Comment (bentian) [inline dom/payments/test/browser_payment_construction.js]: `2nd item's ...`

File: dom/payments/test/browser_payment_construction.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        let modifiers = details.modifiers;
+        is(modifiers.length, 1, "modifiers' length should be 1.");
+        let modifier = modifiers.queryElementAt(0, Ci.nsIPaymentDetailsModifier);
+        let modifierSupportedMethods = modifier.supportedMethods;
+        is(modifierSupportedMethods.length, 1, "modifier's supported methods length should be 1.");
+        supportedMethod = modifierSupportedMethods.queryElementAt(0, Ci.nsISupportsString);
+        is(supportedMethod, "visa", "modifier supported method name should be 'visa'.");
```
Comment (bentian) [inline dom/payments/test/browser_payment_construction.js]: `modifier's supported method name ...`

File: dom/payments/test/browser_payment_construction.js [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        is(modifier.total.label, "Discounted donation", "modifier's total label should be 'Discounted donation'.");
+        is(modifier.total.amount.currency, "USD", "modifier's total currency should be 'USD'.");
+        is(modifier.total.amount.value, "45.00", "modifier's total value should be '45.00'.");
+        let additionalItems = modifier.additionalDisplayItems;
+        is(additionalItems.length, "1", "additionalDisplayItems length should be 1.");
```
Comment (bentian) [inline dom/payments/test/browser_payment_construction.js]: `additionalDisplayItems'`

File: dom/payments/test/browser_payment_construction.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        let additionalItem = additionalItems.queryElementAt(0, Ci.nsIPaymentItem);
+        is(additionalItem.label, "VISA discount", "additional item label should be 'VISA discount'.");
+        is(additionalItem.amount.currency, "USD", "additional item currency should be 'USD'.");
+        is(additionalItem.amount.value, "-10.00", "additional item value should be '-10.00'.");
+        is(modifier.data, "{\"discountProgramParticipantId\":\"86328764873265\"}",
+           "the modifier's data should be '{\"discountProgramParticipantId\":\"86328764873265\"}'.");
```
Comment (bentian) [inline dom/payments/test/browser_payment_construction.js]: Remove `the`.

File: dom/payments/test/browser_payment_construction.js [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        is(supportedMethod, "visa", "modifier supported method name should be 'visa'.");
+        is(modifier.total.label, "Discounted donation", "modifier's total label should be 'Discounted donation'.");
+        is(modifier.total.amount.currency, "USD", "modifier's total currency should be 'USD'.");
+        is(modifier.total.amount.value, "45.00", "modifier's total value should be '45.00'.");
+        let additionalItems = modifier.additionalDisplayItems;
+        is(additionalItems.length, "1", "additionalDisplayItems length should be 1.");
+        let additionalItem = additionalItems.queryElementAt(0, Ci.nsIPaymentItem);
+        is(additionalItem.label, "VISA discount", "additional item label should be 'VISA discount'.");
+        is(additionalItem.amount.currency, "USD", "additional item currency should be 'USD'.");
+        is(additionalItem.amount.value, "-10.00", "additional item value should be '-10.00'.");
```
Comment (bentian) [inline dom/payments/test/browser_payment_construction.js]: `additional item's ...` for above.

File: dom/payments/test/browser_payment_construction.js [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+      ok(paymentEnum.hasMoreElements(), "Should have at least one request in nsIPaymentRequestService.");
+      while (paymentEnum.hasMoreElements()) {
+        let payment = paymentEnum.getNext().QueryInterface(Ci.nsIPaymentRequest);
+        ok(payment, "Should get existing payment.");
+
+        // checking the passed PaymentMethods parameter
+        is(payment.paymentMethods.length, 1, "Should only one supported payment method.");
+        let paymentMethod = payment.paymentMethods.queryElementAt(0, Ci.nsIPaymentMethodData);
+        ok(paymentMethod, "Get payment method failed.");
+        is(paymentMethod.supportedMethods.length, 2, "supportedMethods length should be 2.");
```
Comment (bentian) [inline dom/payments/test/browser_payment_construction.js]: `supportedMethods'`

File: dom/payments/test/browser_payment_construction.js [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+      ok(paymentSrv, "Get payment request service failed.");
+      let paymentEnum = paymentSrv.enumerate();
+      ok(paymentEnum.hasMoreElements(), "Should have at least one request in nsIPaymentRequestService.");
+      while (paymentEnum.hasMoreElements()) {
+        let payment = paymentEnum.getNext().QueryInterface(Ci.nsIPaymentRequest);
+        ok(payment, "Should get existing payment.");
+
+        // checking the passed PaymentMethods parameter
+        is(payment.paymentMethods.length, 1, "Should only one supported payment method.");
+        let paymentMethod = payment.paymentMethods.queryElementAt(0, Ci.nsIPaymentMethodData);
```
Comment (bentian) [inline dom/payments/test/browser_payment_construction.js]: Rename to `methodData`.

File: dom/payments/test/browser_payment_construction.js [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+const kTestPage = kTestRoot + "payment_request.html";
+
+add_task(function*() {
+  yield BrowserTestUtils.withNewTab(kTestPage,
+    function*(browser) {
+      let paymentSrv = Cc["@mozilla.org/dom/payments/payment-request-service;1"].getService(Ci.nsIPaymentRequestService);
+      ok(paymentSrv, "Get payment request service failed.");
+      let paymentEnum = paymentSrv.enumerate();
+      ok(paymentEnum.hasMoreElements(), "Should have at least one request in nsIPaymentRequestService.");
```
Comment (bentian) [inline dom/payments/test/browser_payment_construction.js]: `nsIPaymentRequestService should have at least one request`.

File: dom/payments/test/browser_payment_construction.js (hunk 1 above)
Comment (bentian) [inline dom/payments/test/browser_payment_construction.js]: `Should be only one payment method`.

File: dom/payments/test/browser_payment_construction.js [hunk 12]
```diff
@@ -0,0 +1,72 @@
+"use strict";
+
+const kTestPage = kTestRoot + "payment_request.html";
+
+add_task(function*() {
+  yield BrowserTestUtils.withNewTab(kTestPage,
+    function*(browser) {
+      let paymentSrv = Cc["@mozilla.org/dom/payments/payment-request-service;1"].getService(Ci.nsIPaymentRequestService);
```
Comment (marcoscaceres) [inline dom/payments/test/browser_payment_construction.js]: nit: s/let/const/ (same for any variable that is not reassigned)

File: dom/payments/test/browser_payment_construction.js [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+        let payment = paymentEnum.getNext().QueryInterface(Ci.nsIPaymentRequest);
+        ok(payment, "Should get existing payment.");
+
+        // checking the passed PaymentMethods parameter
+        is(payment.paymentMethods.length, 1, "Should only one supported payment method.");
+        let paymentMethod = payment.paymentMethods.queryElementAt(0, Ci.nsIPaymentMethodData);
+        ok(paymentMethod, "Get payment method failed.");
+        is(paymentMethod.supportedMethods.length, 2, "supportedMethods length should be 2.");
+        let supportedMethod = paymentMethod.supportedMethods.queryElementAt(0, Ci.nsISupportsString);
+        is(supportedMethod, "visa", "payment method should be 'visa'.");
```
Comment (marcoscaceres) [inline dom/payments/test/browser_payment_construction.js]: I'm not comfortable we are testing against potentially real payment methods. We should probably use fake names here.

File: dom/payments/test/browser_payment_construction.js [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+        supportedMethod = paymentMethod.supportedMethods.queryElementAt(1, Ci.nsISupportsString);
+        is(supportedMethod, "mastercard", "payment method should be 'mastercard'.");
```
Comment (marcoscaceres) [inline dom/payments/test/browser_payment_construction.js]: As above.

File: dom/payments/test/browser_payment_construction.js [hunk 15]
```diff
@@ -0,0 +1,72 @@
+"use strict";
+
+const kTestPage = kTestRoot + "payment_request.html";
```
Comment (marcoscaceres) [inline dom/payments/test/browser_payment_construction.js]: Maybe add add a comment that kTestPage is defined in dom/payments/test/head.js

File: dom/payments/test/payment_request.html [hunk 16]
```diff
@@ -0,0 +1,49 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <title>Payment Request Testing</title>
+    <meta content="text/html;charset=utf-8" http-equiv="Content-Type">
+    <meta content="utf-8" http-equiv="encoding">
+  </head>
+  <body>
+    <script type="text/javascript">
+    var supportedInstruments = [{
```
Comment (marcoscaceres) [inline dom/payments/test/payment_request.html]: s/var/const/g

File: dom/payments/test/payment_request.html [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+      supportedMethods: [ 'visa', 'mastercard', ]
```
Comment (marcoscaceres) [inline dom/payments/test/payment_request.html]: Nit: use double quotes globally for strings.

**Pull request closed without merging**